src/cli_util.rs

```diff
@@ -987,21 +991,82 @@ impl WorkspaceCommandHelper {
     }
 
     pub fn parse_revset(
         &self,
         revision_str: &str,
+        ui: Option<&mut Ui>,
     ) -> Result<Rc<RevsetExpression>, RevsetParseError> {
         let expression = revset::parse(
             revision_str,
             &self.revset_aliases_map,
             Some(&self.revset_context()),
         )?;
+        if let Some(ui) = ui {
+            fn has_legacy_rule(expression: &Rc<RevsetExpression>) -> bool {
+                match expression.as_ref() {
+                    RevsetExpression::None => false,
+                    RevsetExpression::All => false,
+                    RevsetExpression::Commits(_) => false,
+                    RevsetExpression::CommitRef(_) => false,
+                    RevsetExpression::Ancestors {
+                        heads,
+                        generation: _,
+                        is_legacy,
+                    } => *is_legacy || has_legacy_rule(heads),
+                    RevsetExpression::Descendants {
+                        roots,
+                        generation: _,
+                        is_legacy,
+                    } => *is_legacy || has_legacy_rule(roots),
+                    RevsetExpression::Range {
+                        roots,
+                        heads,
+                        generation: _,
+                    } => has_legacy_rule(roots) || has_legacy_rule(heads),
+                    RevsetExpression::DagRange {
+                        roots,
+                        heads,
+                        is_legacy,
+                    } => *is_legacy || has_legacy_rule(roots) || has_legacy_rule(heads),
+                    RevsetExpression::Heads(expression) => has_legacy_rule(expression),
+                    RevsetExpression::Roots(expression) => has_legacy_rule(expression),
+                    RevsetExpression::Latest {
+                        candidates,
+                        count: _,
+                    } => has_legacy_rule(candidates),
+                    RevsetExpression::Filter(_) => false,
+                    RevsetExpression::AsFilter(expression) => has_legacy_rule(expression),
+                    RevsetExpression::Present(expression) => has_legacy_rule(expression),
+                    RevsetExpression::NotIn(expression) => has_legacy_rule(expression),
+                    RevsetExpression::Union(expression1, expression2) => {
+                        has_legacy_rule(expression1) || has_legacy_rule(expression2)
+                    }
+                    RevsetExpression::Intersection(expression1, expression2) => {
+                        has_legacy_rule(expression1) || has_legacy_rule(expression2)
+                    }
+                    RevsetExpression::Difference(expression1, expression2) => {
+                        has_legacy_rule(expression1) || has_legacy_rule(expression2)
+                    }
+                }
+            }
+            if has_legacy_rule(&expression) {
```

Comment:
Another way around is to pass `allow_legacy` flag to parser.
```rust
let r = parse(..., allow_legacy = false);
if matches!(r, Err(...)) {
    parse(..., allow_legacy = true)?;
    writeln!(ui.warning(), ...);
}
```

I guess it would be slightly simpler, but I don't have strong opinion. This is just a temporary hack and we can remove it anyway.

Reply:
Yeah, that might have been simpler. Since it's already done now and this code doesn't change that much, I'll just leave it as is.